### PR TITLE
Fixed inconsistent return types

### PR DESCRIPTION
**Pull request by gadelat**

### CODE REVIEW

File: system/database/DB_utility.php [hunk 1]
```diff
@@ -316,7 +316,7 @@ public function xml_from_result($query, $params = array())
 	 * Database Backup
 	 *
 	 * @param	array	$params
-	 * @return	void
+	 * @return	mixed
```
Comment (narfbg) [inline system/database/DB_utility.php]: If anything, this should be string.

File: system/helpers/download_helper.php [hunk 2]
```diff
@@ -59,7 +59,7 @@
 	 * @param	string	filename
 	 * @param	mixed	the data to be downloaded
 	 * @param	bool	whether to try and send the actual file MIME type
-	 * @return	void
+	 * @return	bool
```
Comment (narfbg) [inline system/helpers/download_helper.php]: No boolean value would make sense for this function ... if it returns such, it should be changed to void.

File: system/libraries/Migration.php [hunk 3]
```diff
@@ -421,7 +421,7 @@ protected function _get_version()
 	 * Stores the current schema version
 	 *
 	 * @param	string	$migration	Migration reached
-	 * @return	void	Outputs a report of the migration
+	 * @return	object	Outputs a report of the migration
```
Comment (narfbg) [inline system/libraries/Migration.php]: It returns the return value of `$this->db->update()`, which is boolean.
The return should be removed, this is meant to be void.

File: user_guide_src/source/database/utilities.rst [hunk 4]
```diff
@@ -252,7 +252,7 @@ Class Reference
 	.. php:method:: backup([$params = array()])
 
 		:param	array	$params: An associative array of options
-		:returns:	void
+		:returns:	mixed
```
Comment (narfbg) [inline user_guide_src/source/database/utilities.rst]: ':returns:' is a description
':rtype:' is the type

**Comment by gadelat** [comment]
Thank you for a review

**Comment by kakysha** [comment]
Also, session database & files drivers read() method returns '', while memchached & redis driver returns FALSE.

File: system/database/DB_utility.php [hunk 5]
```diff
@@ -417,7 +417,7 @@ public function backup($params = array())
 			return gzencode($this->_backup($prefs));
 		}
 
-		return;
+		return '';
```
Comment (narfbg) [inline system/database/DB_utility.php]: Revert this.

File: system/libraries/Session/drivers/Session_files_driver.php [hunk 6]
```diff
@@ -156,13 +156,13 @@ public function read($session_id)
 				if (($this->_file_handle = fopen($this->_file_path.$session_id, 'w+b')) === FALSE)
 				{
 					log_message('error', "Session: File '".$this->_file_path.$session_id."' doesn't exist and cannot be created.");
-					return FALSE;
+					return '';
```
Comment (narfbg) [inline system/libraries/Session/drivers/Session_files_driver.php]: And this.

File: system/libraries/Session/drivers/Session_files_driver.php [hunk 7]
```diff
@@ -170,7 +170,7 @@ public function read($session_id)
 				log_message('error', "Session: Unable to obtain lock for file '".$this->_file_path.$session_id."'.");
 				fclose($this->_file_handle);
 				$this->_file_handle = NULL;
-				return FALSE;
+				return '';
```
Comment (narfbg) [inline system/libraries/Session/drivers/Session_files_driver.php]: And this.

File: system/libraries/Session/drivers/Session_files_driver.php [hunk 8]
```diff
@@ -156,13 +156,13 @@ public function read($session_id)
...
 					log_message('error', "Session: File '".$this->_file_path.$session_id."' doesn't exist and cannot be created.");
-					return FALSE;
+					return '';
 				}
 			}
 			elseif (($this->_file_handle = fopen($this->_file_path.$session_id, 'r+b')) === FALSE)
 			{
 				log_message('error', "Session: Unable to open file '".$this->_file_path.$session_id."'.");
-				return FALSE;
+				return '';
```
Comment (narfbg) [inline system/libraries/Session/drivers/Session_files_driver.php]: And this too.

File: system/libraries/Session/drivers/Session_memcached_driver.php [hunk 9]
```diff
@@ -170,7 +170,7 @@ public function read($session_id)
 			return $session_data;
 		}
 
-		return FALSE;
+		return '';;
```
Comment (narfbg) [inline system/libraries/Session/drivers/Session_memcached_driver.php]: Revert.

File: system/libraries/Session/drivers/Session_redis_driver.php [hunk 10]
```diff
@@ -171,7 +171,7 @@ public function read($session_id)
 			return $session_data;
 		}
 
-		return FALSE;
+		return '';
```
Comment (narfbg) [inline system/libraries/Session/drivers/Session_redis_driver.php]: Revert.

File: user_guide_src/source/database/utilities.rst [hunk 11]
```diff
@@ -252,8 +252,8 @@ Class Reference
 	.. php:method:: backup([$params = array()])
 
 		:param	array	$params: An associative array of options
-		:returns:	void
-		:rtype:	void
+		:returns:	string
```
Comment (narfbg) [inline user_guide_src/source/database/utilities.rst]: As I said, this is supposed to be a description. 'string' is a data type, not a description.
You need to describe what the string represents.

**Comment by narfbg** [comment]
Don't touch the session read() returns, they are all intentional.

**Comment by kakysha** [comment]
@narfbg why then they are inconsistent between drivers? Also, from http://php.net/manual/en/sessionhandlerinterface.read.php
> If the record was not found, return an empty string.

**Comment by narfbg** [comment]
Lack of a record is not an error, failure to open one is.

**Pull request merged**